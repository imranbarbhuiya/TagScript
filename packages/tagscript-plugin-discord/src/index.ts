export type * from './lib/interfaces';
export * from './lib/Transformer';
export * from './lib/Parsers';
export * from './lib/Utils';
