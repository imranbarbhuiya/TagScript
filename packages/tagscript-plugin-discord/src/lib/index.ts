export * from './Interpreter';
export * from './Transformer';
export * from './Parsers';
export * from './interfaces';
export * from './Utils/Util';
