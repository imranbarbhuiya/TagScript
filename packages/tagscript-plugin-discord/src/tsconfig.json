{
	"extends": "../../../tsconfig.json",
	"compilerOptions": {
		"rootDir": "./",
		"outDir": "../dist",
		"incremental": false
	},
	"include": ["."]
}
