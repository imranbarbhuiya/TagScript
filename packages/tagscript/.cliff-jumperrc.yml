name: tagscript
monoRepo: true
packagePath: packages/tagscript
tagTemplate: "{{full-name}}@{{new-version}}"