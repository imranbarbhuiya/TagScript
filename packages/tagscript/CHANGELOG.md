# Changelog

All notable changes to this project will be documented in this file.

# [tagscript@1.3.1](https://github.com/imranbarbhuiya/tagscript/compare/tagscript@1.3.0...tagscript@1.3.1) - (2023-05-27)

## 🐛 Bug Fixes

- Use types first ([8818025](https://github.com/imranbarbhuiya/tagscript/commit/881802549a95f8d3a561246e03e49693cef97305))

## 📝 Documentation

- Update to new links ([7f0f5ed](https://github.com/imranbarbhuiya/tagscript/commit/7f0f5edbd97bc2687db41813666bf39d33613ce4))

# [tagscript@1.3.0](https://github.com/imranbarbhuiya/tagscript/compare/tagscript@1.2.14...tagscript@1.3.0) - (2022-12-09)

## 🐛 Bug Fixes

- Rename files and bump deps ([f73eaa3](https://github.com/imranbarbhuiya/tagscript/commit/f73eaa381a9c14047121160d7acf57834c2890cc))
- Code block copy button ([ead3152](https://github.com/imranbarbhuiya/tagscript/commit/ead31521ac4406b30a6a090b203ba5587743f908))
- Relative url ([6e7b0f9](https://github.com/imranbarbhuiya/tagscript/commit/6e7b0f98cf0cc51357c1a582ad9f736b93223b84))

## 🚀 Features

- Add parsers ([7c5bfe2](https://github.com/imranbarbhuiya/tagscript/commit/7c5bfe28031375bb71c96488376749d8d53c9f33))

# [tagscript@1.2.12](https://github.com/imranbarbhuiya/tagscript/compare/tagscript@1.2.11...tagscript@1.2.12) - (2022-10-10)

## 📝 Documentation

- Add some examples ([453668b](https://github.com/imranbarbhuiya/tagscript/commit/453668b0c5cb86875ed2906085bdbbe898f59cdc))
- Fix examples ([89b6c49](https://github.com/imranbarbhuiya/tagscript/commit/89b6c49d6aac810d3856fa42c81394093ec6d429))
- Add some examples ([14a2cd2](https://github.com/imranbarbhuiya/tagscript/commit/14a2cd2ba51b9807d7a7f5a1fca2fa12e654f9c5))
- Add @allcontributors[bot] as a contributor ([23bb4ef](https://github.com/imranbarbhuiya/tagscript/commit/23bb4ef97a9f0feaf43b764e871c43c2a281242f))
- Escape fix ([16f16f1](https://github.com/imranbarbhuiya/tagscript/commit/16f16f18dcddbf9bf02075e7719936703765a290))
- Fix docs ([9ea716d](https://github.com/imranbarbhuiya/tagscript/commit/9ea716d2b663ccec661c20010ad19dff0cc8329a))

# [tagscript@1.2.11](https://github.com/imranbarbhuiya/tagscript/compare/tagscript@1.2.10...tagscript@1.2.11) - (2022-10-05)

## 📝 Documentation

- Add ShoXcy as a contributor for doc (#111) ([c88b338](https://github.com/imranbarbhuiya/tagscript/commit/c88b33898faef5816d5e8192c140b8d8efef814b))
- Fix changelog ([96dd604](https://github.com/imranbarbhuiya/tagscript/commit/96dd604f2c9e2f1fd2409e26f71848f8a9b829cb))

## 🪞 Styling

- Setup lint again (#114) ([ea7a209](https://github.com/imranbarbhuiya/tagscript/commit/ea7a20956408b006096d585447ec2630cc4285cb))

# [tagscript@1.2.12](https://github.com/imranbarbhuiya/tagscript/compare/tagscript@1.2.11...tagscript@1.2.12) - (2022-10-10)

## 📝 Documentation

- Add some examples ([453668b](https://github.com/imranbarbhuiya/tagscript/commit/453668b0c5cb86875ed2906085bdbbe898f59cdc))
- Fix examples ([89b6c49](https://github.com/imranbarbhuiya/tagscript/commit/89b6c49d6aac810d3856fa42c81394093ec6d429))
- Add some examples ([14a2cd2](https://github.com/imranbarbhuiya/tagscript/commit/14a2cd2ba51b9807d7a7f5a1fca2fa12e654f9c5))
- Add @allcontributors[bot] as a contributor ([23bb4ef](https://github.com/imranbarbhuiya/tagscript/commit/23bb4ef97a9f0feaf43b764e871c43c2a281242f))
- Escape fix ([16f16f1](https://github.com/imranbarbhuiya/tagscript/commit/16f16f18dcddbf9bf02075e7719936703765a290))
- Fix docs ([9ea716d](https://github.com/imranbarbhuiya/tagscript/commit/9ea716d2b663ccec661c20010ad19dff0cc8329a))

# [tagscript@1.2.11](https://github.com/imranbarbhuiya/tagscript/compare/tagscript@1.2.10...tagscript@1.2.11) - (2022-10-05)

## 📝 Documentation

- Add ShoXcy as a contributor for doc (#111) ([c88b338](https://github.com/imranbarbhuiya/tagscript/commit/c88b33898faef5816d5e8192c140b8d8efef814b))
- Fix changelog ([96dd604](https://github.com/imranbarbhuiya/tagscript/commit/96dd604f2c9e2f1fd2409e26f71848f8a9b829cb))

## 🪞 Styling

- Setup lint again (#114) ([ea7a209](https://github.com/imranbarbhuiya/tagscript/commit/ea7a20956408b006096d585447ec2630cc4285cb))

# [tagscript@1.2.12](https://github.com/imranbarbhuiya/tagscript/compare/tagscript@1.2.11...tagscript@1.2.12) - (2022-10-10)

## 📝 Documentation

- Add some examples ([453668b](https://github.com/imranbarbhuiya/tagscript/commit/453668b0c5cb86875ed2906085bdbbe898f59cdc))
- Fix examples ([89b6c49](https://github.com/imranbarbhuiya/tagscript/commit/89b6c49d6aac810d3856fa42c81394093ec6d429))
- Add some examples ([14a2cd2](https://github.com/imranbarbhuiya/tagscript/commit/14a2cd2ba51b9807d7a7f5a1fca2fa12e654f9c5))
- Add @allcontributors[bot] as a contributor ([23bb4ef](https://github.com/imranbarbhuiya/tagscript/commit/23bb4ef97a9f0feaf43b764e871c43c2a281242f))
- Escape fix ([16f16f1](https://github.com/imranbarbhuiya/tagscript/commit/16f16f18dcddbf9bf02075e7719936703765a290))
- Fix docs ([9ea716d](https://github.com/imranbarbhuiya/tagscript/commit/9ea716d2b663ccec661c20010ad19dff0cc8329a))

# [tagscript@1.2.11](https://github.com/imranbarbhuiya/tagscript/compare/tagscript@1.2.10...tagscript@1.2.11) - (2022-10-05)

## 📝 Documentation

- Add ShoXcy as a contributor for doc (#111) ([c88b338](https://github.com/imranbarbhuiya/tagscript/commit/c88b33898faef5816d5e8192c140b8d8efef814b))
- Fix changelog ([96dd604](https://github.com/imranbarbhuiya/tagscript/commit/96dd604f2c9e2f1fd2409e26f71848f8a9b829cb))

## 🪞 Styling

- Setup lint again (#114) ([ea7a209](https://github.com/imranbarbhuiya/tagscript/commit/ea7a20956408b006096d585447ec2630cc4285cb))

# [tagscript@1.2.9](https://github.com/imranbarbhuiya/tagscript/compare/tagscript@1.2.8...tagscript@1.2.9) - (2022-08-21)

## 📝 Documentation

- Improve documentation (#53) ([69c5d6f](https://github.com/imranbarbhuiya/tagscript/commit/69c5d6ff802d9d788959762abb64ae310d4b848a))

# [tagscript@1.2.3](https://github.com/imranbarbhuiya/tagscript/compare/tagscript@1.2.1...tagscript@1.2.3) - (2022-07-06)

## 📝 Documentation

- Add UndiedGamer as a contributor for code (#52) ([6376f48](https://github.com/imranbarbhuiya/tagscript/commit/6376f4826b9e778246a1536d8cd0b327c8000484))

## 🧪 Testing

- Migrate to vitest (#69) ([ed23618](https://github.com/imranbarbhuiya/tagscript/commit/ed23618afed995332b0b0792a6fc7e97cd1867cb))

# [1.2.0](https://github.com/imranbarbhuiya/tagscript/compare/tagscript@1.1.0...v1.2.0) - (2022-05-16)

## 🚀 Features

- Allow sending files (#43) ([f616c5c](https://github.com/imranbarbhuiya/tagscript/commit/f616c5cd66ff61cc47b889820fa818465ae56e3e))

# [1.1.0](https://github.com/imranbarbhuiya/tagscript/compare/tagscript@1.0.2...v1.1.0) - (2022-05-16)

## 🚀 Features

- New json parser (#42) ([ea0cfe1](https://github.com/imranbarbhuiya/tagscript/commit/ea0cfe1e73973c3503c6c7660029c1a8c0f10408))

# [1.0.2]

(<https://github.com/imranbarbhuiya/tagscript/tree/v1.0.2>) - (2022-05-10)

## 🐛 Bug Fixes

- Properly format capitalize (#34) ([9c3b8ee](https://github.com/imranbarbhuiya/tagscript/commit/9c3b8eecb1b55646d0b5536c7380615e430768cd))
- **docs:** Add mdn links (#27) ([82073c3](https://github.com/imranbarbhuiya/tagscript/commit/82073c306d2bdb3e10bbd1328c96ff1a5cdde535))
- Remove browser build from plugin (#25) ([fa5ec4d](https://github.com/imranbarbhuiya/tagscript/commit/fa5ec4dbbc257e93cf5dc0ccba76ed9111c8d9a8))}

### 💥 Breaking Changes

- Prepack scripts (#24) ([48001bb](https://github.com/imranbarbhuiya/tagscript/commit/48001bbeb43c6239d645e6180586655d0aadb560))

## 📝 Documentation

- Add favna as a contributor for maintenance (#29) ([4359983](https://github.com/imranbarbhuiya/tagscript/commit/435998336c0b4ce959046363e3875ab96ccd8384))

## 🚀 Features

- Move all discord.js related parsers into a new package (#22) ([f7e27fa](https://github.com/imranbarbhuiya/tagscript/commit/f7e27fae5a0629679415f0e8c84e3fdfde452411))}

### 💥 Breaking Changes
