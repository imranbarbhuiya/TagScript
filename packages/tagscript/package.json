{
	"name": "tagscript",
	"version": "1.0.1",
	"description": "A simple and safe template engine.",
	"main": "dist/index.js",
	"module": "dist/index.mjs",
	"browser": "dist/index.global.js",
	"unpkg": "dist/index.global.js",
	"types": "dist/index.d.ts",
	"exports": {
		"import": "./dist/index.mjs",
		"require": "./dist/index.js",
		"types": "./dist/index.d.ts"
	},
	"sideEffects": false,
	"author": "@imranbarbhuiya",
	"scripts": {
		"lint": "eslint src --ext ts --fix -c ../../.eslintrc",
		"build": "tsup && tsc -b src",
		"prepack": "rollup-type-bundler",
		"bump": "cliff-jumper",
		"check-update": "cliff-jumper --dry-run",
		"test": "jest"
	},
	"keywords": [
		"tagscript",
		"string parser",
		"safe string",
		"typescript",
		"template engine",
		"template",
		"template string",
		"tag"
	],
	"devDependencies": {
		"@favware/rollup-type-bundler": "^1.0.7",
		"tsup": "^5.12.7",
		"typescript": "^4.6.4"
	},
	"engines": {
		"node": ">=v14.0.0"
	},
	"files": [
		"dist",
		"!dist/*.tsbuildinfo"
	],
	"repository": {
		"type": "git",
		"url": "git+https://github.com/imranbarbhuiya/tagscript.git",
		"directory": "packages/tagscript"
	},
	"bugs": {
		"url": "https://github.com/imranbarbhuiya/tagscript/issues"
	},
	"homepage": "https://tagscript.js.org/"
}
