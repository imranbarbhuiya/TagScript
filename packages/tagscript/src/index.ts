export * from './lib/Interpreter';
export * from './lib/Transformer';
export * from './lib/Parsers';
export type * from './lib/interfaces';
export * from './lib/Utils/Util';
