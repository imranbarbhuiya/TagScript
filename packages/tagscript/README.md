<div align="center">

# TagScript

**A simple and safe template engine.**

[![npm](https://img.shields.io/npm/dw/tagscript)](https://www.npmjs.com/package/tagscript)
[![codecov](https://codecov.io/gh/imranbarbhuiya/tagscript/branch/main/graph/badge.svg)](https://codecov.io/gh/imranbarbhuiya/tagscript)
[![npm](https://img.shields.io/npm/v/tagscript?color=crimson&logo=npm&style=flat-square)](https://www.npmjs.com/package/tagscript)

</div>

## Description

TagScript is a drop in easy to use string interpreter that lets you provide users with ways of customizing their profiles or chat rooms with interactive text.

Read Full Documentation [here](https://tagscript.js.org/).

---

## Features

-   Written In Typescript
-   Offers CJS, ESM and UMD builds
-   Full TypeScript & JavaScript support
-   Faster than âš¡
-   Simple, expressive and safe template engine.
-   Supports your own parsers and transformers

## Installation

```bash
# npm
npm i tagscript

# yarn
yarn add tagscript

```

## Usage

---

**Note:** While examples uses `import`, it maps 1:1 with CommonJS' require syntax. For example,

```ts
import { Interpreter } from 'tagscript';
```

is the same as

```js
const { Interpreter } = require('tagscript');
```

---

```ts
import { Interpreter, RandomParser, RangeParser, FiftyFiftyParser, IfStatementParser, SliceParser } from 'tagscript';
const ts = new Interpreter(new SliceParser(), new FiftyFiftyParser(), new RandomParser(), new IfStatementParser());

const result = await ts.run(
	`
    {random: Parbez,Rkn,Priyansh} attempts to pick the lock!,
    I pick {if({5050:.}!=):heads|tails}
    `
);
```

## Parsers

Parsers are used to parse a tag and return a value based on the tag. You can use our [builtin parsers](https://github.com/imranbarbhuiya/TagScript/tree/main/src/lib/Parsers) or write your own parsers.
Your own parser should implement [IParser](https://tagscript.js.org/interfaces/IParser.html) interface.

```ts
import { Interpreter, RandomParser, RangeParser, FiftyFiftyParser, IfStatementParser, SliceParser } from 'tagscript';
const ts = new Interpreter(new SliceParser(), new FiftyFiftyParser(), new RandomParser(), new IfStatementParser());
```

## Transformers

Transformers are used to transform a value based on the tag at runtime. You can use our [builtin transformers](https://github.com/imranbarbhuiya/TagScript/tree/main/src/lib/Transformer) or write your own transformers.
Your own transformer should implement [ITransformer](https://tagscript.js.org/interfaces/ITransformer.html) interface.

```ts
import { Interpreter, StringTransformer } from 'tagscript';
const ts = new Interpreter();

await ts.run(str, { args: new StringTransformer(args) });
```

## Buy me some doughnuts

If you want to support me by donating, you can do so by using any of the following methods. Thank you very much in advance!

<a href="https://www.buymeacoffee.com/parbez" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/default-orange.png" alt="Buy Me A Coffee" height="41" width="174"></a>
<a href='https://ko-fi.com/Y8Y1CBIJH' target='_blank'><img height='36' style='border:0px;height:36px;' src='https://cdn.ko-fi.com/cdn/kofi4.png?v=3' border='0' alt='Buy Me a Coffee at ko-fi.com' /></a>

## Contributors âœ¨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://github.com/imranbarbhuiya"><img src="https://avatars.githubusercontent.com/u/74945038?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Parbez</b></sub></a><br /><a href="https://github.com/imranbarbhuiya/TagScript/commits?author=imranbarbhuiya" title="Code">ðŸ’»</a> <a href="#maintenance-imranbarbhuiya" title="Maintenance">ðŸš§</a> <a href="#ideas-imranbarbhuiya" title="Ideas, Planning, & Feedback">ðŸ¤”</a></td>
    <td align="center"><a href="https://renovate.whitesourcesoftware.com"><img src="https://avatars.githubusercontent.com/u/25180681?v=4?s=100" width="100px;" alt=""/><br /><sub><b>WhiteSource Renovate</b></sub></a><br /><a href="#maintenance-renovate-bot" title="Maintenance">ðŸš§</a></td>
    <td align="center"><a href="https://favware.tech"><img src="https://avatars.githubusercontent.com/u/4019718?v=4?s=100" width="100px;" alt=""/><br /><sub><b>Jeroen Claassens</b></sub></a><br /><a href="#maintenance-favna" title="Maintenance">ðŸš§</a></td>
  </tr>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

## Special Thanks

-   [JonSnowbd](https://github.com/JonSnowbd/) for creating TagScript in python.
