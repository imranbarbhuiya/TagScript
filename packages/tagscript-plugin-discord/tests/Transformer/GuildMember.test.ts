import { Interpreter, StrictVarsParser } from 'tagscript';
import { MemberTransformer } from '../../src';
import { member } from './Structures';

const ts = new Interpreter(new StrictVarsParser());

describe('MemberTransformer', () => {
	test('GIVEN a member tag THEN return value from member variable', async () => {
		expect((await ts.run('{member}', { member: new MemberTransformer(member) })).body).toBe('<@758880890159235083>');
		expect((await ts.run('{member(nickname)}', { member: new MemberTransformer(member) })).body).toBe('');
	});
});
