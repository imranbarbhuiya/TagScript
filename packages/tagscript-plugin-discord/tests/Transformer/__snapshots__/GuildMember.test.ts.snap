// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MemberTransformer > should match the snapshot 1`] = `
{
  "avatar": null,
  "bot": false,
  "color": "",
  "createdAt": "2020-09-25T02:41:43.539Z",
  "createdTimestamp": 1601001703539,
  "discriminator": "0000",
  "displayAvatar": "https://cdn.discordapp.com/avatars/758880890159235083/17ac5f89d5f8b08b5bbd6cc43c930399.webp",
  "displayName": "Parbez",
  "id": "758880890159235083",
  "joinedAt": "2022-01-19T16:52:53.953Z",
  "joinedTimestamp": 1642611173953,
  "mention": "<@758880890159235083>",
  "name": "",
  "nickname": null,
  "position": 1,
  "roleIds": "933378013154906142, 933368398996447292",
  "roleNames": "., @everyone",
  "roles": "<@&933378013154906142> @everyone",
  "tag": "parbez#0000",
  "timeoutUntil": "",
  "timeoutUntilTimestamp": null,
  "topRole": ".",
  "username": "parbez",
}
`;
